### DISCUSSION TITLE
Does opentelemetry support traceId propagation with websockets?

### DISCUSSION THREAD

**Discussion by pwfarwell** [open-telemetry/opentelemetry-java#3345] [Q&A]
Our application uses the spring-websocket library.
We have integrated with manual otel tracer support (not the automated java-agent)

If this is supported in otel 1.1 (or higher), is the header injection pattern the same as standard http headers? 
Or is there a different approach I need with websockets?

(Apologies in advance - I searched for 'opentelemetry support for websockets', but couldn't find much. Perhaps that is the answer :) )

**Comment by jkwatson** [comment]
Hi @pwfarwell . I don't know of anyone who has done this yet. Are you looking to propagate trace/span ids over the websocket connection, on a per-message basis, or just have a single traceid/spanid propagated on the opening of the connection?

As far as I know, the per-message propagation would certainly have to be something custom, depending on the specific protocol that's being used over the websocket, but if you had a way to write a Getter/Setter implementation that would write the keys into something in your message structure, then I would guess our existing propagators could work.

**Reply by pwfarwell** [reply]
Thanks @jkwatson The second option might be good enough:

> or just have a single traceid/spanid propagated on the opening of the connection?

I'm guessing that on the initial connection we can include the trace headers, similar to how we inject headers into a regular HTTP request.

**Reply by jkwatson** [reply]
From an instrumentation perspective, then, this is just doing the same instrumentation that you might do for a "normal" http call, correct? You can just use the standard propagators and set headers using a custom Setter for the specific client that you're using (assuming your websocket client gives you access to request headers!).

**Reply by pwfarwell** [reply]
Yes @jkwatson. That is what we will try. Thanks!

**Reply by piyush-harness** [reply]
@pwfarwell Did you get this working, I am exploring a similar usecase for websocket created using Atmosphere lib.

**Reply by cforce** [reply]
is there any reusable solution   meanwhile?
